### ISSUE TITLE
PresentationService does not work with Android 7.1

### ISSUE THREAD

**Issue by ashughes** [commonsguy/cwac-presentation#9]
On Android 7.1, content displayed using the PresentationService method works only for a few seconds, then the content fades out and it stops working.

Tested the presentation service demo on:

Android 7.1 device and emulator - both did not work
Android 7.0 emulator - worked correctly

So it appears that there was a change in Android 7.1 that's causing this to not work anymore.

The most puzzling part is that it _does_ work for a few seconds, then fades out and stops working. As far as I can tell, none of the DisplayListener callbacks are being called when this occurs, so there's no way of even knowing that it has stopped working (from the app's point of view).

I'm wondering if the WindowManager.LayoutParams type and/or flags need tweaking to work on Android 7.1, or if this is simply a bug in Android 7.1.

**Comment by commonsguy** [comment]
What are you using to test this library on Android 7.1? I am not aware of any hardware that runs 7.1 and supports external displays.

Your idea of the cause of the problem is probably spot-on (many thanks!), but I cannot work out a fix without a test environment.

**Comment by ashughes** [comment]
Sorry, should have clarified. You can use the *Simulate secondary displays* option in the developer settings for testing.

In the real world, this is reproducible using Chromecast remote display and Android 7.1. That's how we discovered the issue.

**Comment by commonsguy** [comment]
OK, I think I'm seeing the same thing. If you run the `demoService` sample app, you see the initial slide for a few seconds, then it fades away, and the regular screen mirroring takes over. Is that what you are seeing?

**Comment by ashughes** [comment]
Yes, that's exactly what I'm seeing.

It totally makes sense now after reading Android bug I posted earlier. Since we're using the TYPE_TOAST WindowManager.LayoutParam, it's now behaving like a Toast (fading out after a few seconds). Unfortunately, I'm not sure what other type we _can_ use...

I submitted a feature request for a "proper" way to do this moving forward (http://b.android.com/261681), but if you can figure out a way to do it on Android 7.1, that would be greatly appreciated.

**Comment by commonsguy** [comment]
`TYPE_PHONE` works but requires the `SYSTEM_ALERT_WINDOW` permission. Nothing else seems to work without that permission. However, I am not going to ship an update that requires that permission, as I find that Google's handling of that permission to be vile.

For the time being, I suggest that you fork the project, and switch to `TYPE_PHONE` for Android 7.1:

```
  protected WindowManager.LayoutParams buildLayoutParams() {
    int windowType=
      Build.VERSION.SDK_INT<=Build.VERSION_CODES.N ?
        WindowManager.LayoutParams.TYPE_TOAST :
        WindowManager.LayoutParams.TYPE_WALLPAPER;

    return(new WindowManager.LayoutParams(
                                          WindowManager.LayoutParams.MATCH_PARENT,
                                          WindowManager.LayoutParams.MATCH_PARENT,
                                          0,
                                          0,
                                          windowType,
                                          0, PixelFormat.OPAQUE));
  }
```

Your activity that starts the service will need to call `Settings.canDrawOverlays(this)` on Android 7.1 and, if that returns `false`, call `startActivity(new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION));`.

In a month or so (when I can next spend time with this), I will run some tests with Android O. O deprecates `TYPE_PHONE` *and* `TYPE_TOAST`. If I can come up with an O approach that does not require `SYSTEM_ALERT_WINDOW`, then I'll ship an update that only uses `TYPE_PHONE` on 7.1 and does something else for Android O. If I cannot find a permission-less solution for Android O &mdash; and if I were a betting man, that's how I'd bet &mdash; I will officially deprecate `PresentationService`.

Frankly, it would not shock me if this whole area goes away with O and forward, given their "multi-display" stuff (throwing activities on separate displays).

**Comment by ashughes** [comment]
Thanks for investigating!

I found that `TYPE_SYSTEM_ALERT` also works in conjunction with the `SYSTEM_ALERT_WINDOW` permission. I'm not sure which type would be better, but both seem to work.

FYI, in your sample code above you used `TYPE_WALLPAPER`, which in my testing did not work.

I agree with what you said about how Google handles the `SYSTEM_ALERT_WINDOW` permission. It's very annoying that this broke in 7.1. I'd be much more okay with it breaking in Android O, where we could then replace the functionality with multi-display support and show an Activity on the secondary display in order to get the same result without requiring a permission that really doesn't have anything to do with what we're doing.

It looks like I can use the following to only require the permission on API 23+:

    <uses-permission-sdk-23 android:name="android.permission.SYSTEM_ALERT_WINDOW"/>

We only need it on API 25, but at least this way it won't bother people below API 23.

Thanks for the details regarding `Settings.canDrawOverlays(this)`. I wasn't aware of that. At least it appears to work by default, and the user would have to disable the ability to draw over other apps to prevent it from working.

**Comment by commonsguy** [comment]
> FYI, in your sample code above you used TYPE_WALLPAPER, which in my testing did not work.

Sorry, that was from my last experiment.

> It looks like I can use the following to only require the permission on API 23+:

Yes.

> At least it appears to work by default, and the user would have to disable the ability to draw over other apps to prevent it from working.

That may depend on how your app is installed. It is enabled by default for development, and it is enabled by default for apps distributed via the Play Store. I do not know what Android does for apps distributed by other channels.

**Comment by ashughes** [comment]
> That may depend on how your app is installed. It is enabled by default for development, and it is enabled by default for apps distributed via the Play Store. I do not know what Android does for apps distributed by other channels.

Actually, it looks like it has to do with the `tarketSdkVersion`:

If it's below 23, then draw over apps is enabled automatically.

After I changed to 23 or higher, the app crashes with the same error as if I didn't even have the `SYSTEM_ALERT_WINDOW` permission. Calling `Settings.canDrawOverlays(this)` and `startActivity(new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION));` allows the user to enable drawing over apps, and then it works.

**Comment by ashughes** [comment]
Well, FWIW, I found the commit that changed how TYPE_TOAST works: https://android.googlesource.com/platform/frameworks/base/+/dc24f93%5E!/

Unfortunately, I don't see any way around it other than what we already discussed.

**Comment by commonsguy** [comment]
I can see why they made the change. It would have been handy had they waited until O for it.

Thanks a ton for your research on this!

**Comment by ashughes** [comment]
Very much agree. :)

**Comment by ashughes** [comment]
It turns out that `TYPE_SYSTEM_ALERT` _does not_ work when presenting to a Chromecast using [Cast Remote Display](https://developers.google.com/cast/docs/remote).

However, I discovered that `TYPE_PRIVATE_PRESENTATION` _does_ work, since a Cast Remote Display display is both a `Display.FLAG_PRIVATE` and `Display.FLAG_PRESENTATION` (from `Display.getFlags()`).

According to the JavaDoc for `TYPE_PRIVATE_PRESENTATION`, another requirement is that the display must also be virtual. I can see that a Cast Remote Display display is virtual by debugging and inspecting the fields, but both the type `Display.TYPE_VIRTUAL` and `Display.getType()` are annotated with `@hide`, which is annoying.

I'm pretty sure that most (if not 100%) of our users on Android 7.1 are using our presentation feature with a Chromecast, so this should fix the issue for us at least for now (and without requiring a confusing permission!).

I'm not sure how other real-world secondary displays will behave on Android 7.1. I don't have any Android 7.1 Miracast or HDMI capable devices (are there any?). However, I assume they would work the same way the simulated secondary display works, which would require the permission. I did inspect Miracast on an Android 5.1.1 device, and it was a non-private presentation display with `TYPE_WIFI`.

**Comment by commonsguy** [comment]
> I'm pretty sure that most (if not 100%) of our users on Android 7.1 are using our presentation feature with a Chromecast, so this should fix the issue for us at least for now (and without requiring a confusing permission!). 

`TYPE_PRIVATE_PRESENTATION` did not work with a Chromecast, using its mirroring option, in my testing yesterday. I did not try it with `SYSTEM_ALERT_WINDOW`.

> I don't have any Android 7.1 Miracast or HDMI capable devices (are there any?).

Not that I am aware of, which is why I asked what you were using in my first comment. :-)

> However, I assume they would work the same way the simulated secondary display works, which would require the permission.

Agreed. Leastways, that's how it was prior to 7.1, and I see no evidence that the `TYPE_TOAST` changes would somehow have relaxed that restriction.

**Comment by ashughes** [comment]
I forgot you could mirror to a Chromecast from the Cast quick setting. You are correct that this does not work with `TYPE_PRIVATE_PRESENTATION`. It acts like a normal secondary display and requires `TYPE_SYSTEM_ALERT` and `SYSTEM_ALERT_WINDOW` permission on Android 7.1.

**Comment by commonsguy** [comment]
FYI, on Android O, multi-display support handles the `PresentationService` scenario. You can have an activity on the external display, where its contents are controlled by something else (e.g., `Notification` actions).

So, my plan is to incorporate one of our discussed hacks for Android 7.1, but steer developers to use the Android O multi-display support for compatible devices.

**Comment by commonsguy** [comment]
OK, I have changed `PresentationService` to use `TYPE_SYSTEM_ALERT` on Android 7.1+ by default. In your case, you will need to override `getWindowType()` when you are using Cast Remote Display and return `TYPE_PRESENTATION`... I think. I have no Cast Remote Display code, so I have no idea if that works, though based on your comments, it should. Let me know if you run into any problems. Thanks for your help!

Addressed in v0.5.0.
